### PR TITLE
fix(python, rust): expr parsing date/timestamp

### PR DESCRIPTION
**Pull request by ion-elgreco**
# Description
We weren't parsing all scalar values yet, parses date32/64 and timestampmicros now as well.

# Related Issue(s)
- fixes https://github.com/delta-io/delta-rs/issues/2344

### CODE REVIEW

File: crates/core/src/delta_datafusion/expr.rs [hunk 1]
```diff
@@ -339,6 +343,36 @@ impl<'a> fmt::Display for ScalarValueFormat<'a> {
             ScalarValue::UInt16(e) => format_option!(f, e)?,
             ScalarValue::UInt32(e) => format_option!(f, e)?,
             ScalarValue::UInt64(e) => format_option!(f, e)?,
+            ScalarValue::Date32(e) => match e {
+                Some(e) => {
+                    let dt =
+                        NaiveDate::from_num_days_from_ce_opt((EPOCH_DAYS_FROM_CE + (*e)).into())
+                            .unwrap();
```
Comment (hntd187) [inline crates/core/src/delta_datafusion/expr.rs:350]: Maybe do a `.map_err(...)?` so we can capture errors here?

File: crates/core/src/delta_datafusion/expr.rs [hunk 2]
```diff
@@ -339,6 +343,36 @@ impl<'a> fmt::Display for ScalarValueFormat<'a> {
...
+                    write!(f, "{}", dt)?
+                }
+                None => write!(f, "NULL")?,
+            },
+            ScalarValue::Date64(e) => match e {
+                Some(e) => write!(
+                    f,
+                    "{}",
+                    NaiveDateTime::from_timestamp_millis((*e).into())
+                        .unwrap()
```
Comment (hntd187) [inline crates/core/src/delta_datafusion/expr.rs:360]: and here

File: crates/core/src/delta_datafusion/expr.rs [hunk 3]
```diff
@@ -339,6 +343,36 @@ impl<'a> fmt::Display for ScalarValueFormat<'a> {
...
+                        .date()
+                )?,
+                None => write!(f, "NULL")?,
+            },
+            ScalarValue::TimestampMicrosecond(e, tz) => match e {
+                Some(e) => match tz {
+                    Some(tz) => write!(
+                        f,
+                        "{}",
+                        NaiveDateTime::from_timestamp_micros(*e).unwrap().and_utc()
```
Comment (hntd187) [inline crates/core/src/delta_datafusion/expr.rs:370]: and here :-)

File: crates/core/src/delta_datafusion/expr.rs [hunk 4]
```diff
@@ -339,6 +343,36 @@ impl<'a> fmt::Display for ScalarValueFormat<'a> {
...
+                    )?,
+                    None => write!(f, "{}", NaiveDateTime::from_timestamp_micros(*e).unwrap())?,
```
Comment (hntd187) [inline crates/core/src/delta_datafusion/expr.rs:372]: and finally here :-)

Comment (ion-elgreco) [reply crates/core/src/delta_datafusion/expr.rs:372]: @hntd187 it actually is not an error but it returns Option<NaiveDateTime> when the value that is passed is out of range.. This generally shouldn't happen but I can return an error when it's None though.

**Review by Blajda — COMMENTED** [review]
The existing unit tests should be extended to ensure we can round trip these new types. https://github.com/delta-io/delta-rs/pull/2357/files#diff-8e1a60f799cd1ad29923ee75871099a19c15ddfc00bef4617d82bdc02cbe3198R523-R526

**Comment by ion-elgreco** [comment]
@Blajda that datafusion sql parser is for some reason parsing a date/timestamp as a substraction : S I also tried wrapping it in single quotes but then it gets parsed as a string. So I am not sure what format datafusion expects datetimes to be

```rust
 left: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "_timestamp_ntz" }), op: Gt, right: Literal(TimestampMicrosecond(1262304000000000, None)) })
 right: BinaryExpr(BinaryExpr { left: Column(Column { relation: None, name: "_timestamp_ntz" }), op: Gt, right: BinaryExpr(BinaryExpr { left: BinaryExpr(BinaryExpr { left: Literal(Int64(2010)), op: Minus, right: Literal(Int64(1)) }), op: Minus, right: Literal(Int64(1)) }) })
```

**Comment by Blajda** [comment]
Yeah you can't dump it directly. The query dialect is postgress so you if you are using a string you can try the cast notation i.e '::date'. Alternative approach use [the date functions](https://arrow.apache.org/datafusion/user-guide/sql/scalar_functions.html#to-timestamp-micros) and convert the scalars to their correct representations.

**Comment by ion-elgreco** [comment]
@Blajda yeah but then Im not really testing round tripping anymore. 

Im assuming the string format needs to be in the same format as the serialization format for the conflict resolution across other engine writers

**Comment by Blajda** [comment]
There's no specification on the query format for conflict resolution with other engines. At least for our engine any queries we convert to a string we should be able to read it back without error

**Review by Blajda — APPROVED** [review]

**Pull request merged**